infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -287,18 +295,30 @@ private void flushCache(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r
     private void rangeWrite(CorfuPayloadMsg<RangeWriteMsg> msg,
                                   ChannelHandlerContext ctx, IServerRouter r) {
         List<LogData> entries = msg.getPayload().getEntries();
-        batchWriter.bulkWrite(entries);
+        batchWriter.bulkWrite(entries, msg.getEpoch());
         r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
     }
 
     /**
-     * Resets the log unit server.
+     * Resets the log unit server via the BatchWriter.
      * Warning: Clears all data.
      */
     @ServerHandler(type = CorfuMsgType.RESET_LOGUNIT)
-    private void resetLogUnit(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        streamLog.reset();
-        dataCache.invalidateAll();
+    private synchronized void resetLogUnit(CorfuPayloadMsg<Long> msg,
+                                           ChannelHandlerContext ctx, IServerRouter r) {
+        // Check if the reset request is with an epoch greater than the last reset epoch seen.
+        // and should be equal to the current router epoch to prevent stale reset requests from
+        // wiping out the data.
+        if (msg.getPayload() > serverContext.getLogUnitBarrierEpoch()
+                && msg.getPayload() == serverContext.getServerEpoch()) {
+            serverContext.setLogUnitBarrierEpoch(msg.getPayload());
+            batchWriter.waitForBarrier(msg.getPayload());
+            batchWriter.reset(msg.getPayload());
+            dataCache.invalidateAll();
+            log.info("LogUnit Server Reset.");
+        } else {
```

Comment:
I wonder if you need to set the "barrier" to Integer.max, install a new cache with the same stream log, call reset, set the "barrier" to the new epoch.